dev/release.sh

```diff
@@ -235,42 +236,58 @@ echo "== Initializing work tree"
 get_version
 
 # Generate a cloned directory name
-clone_branch="openssl-$SERIES.x"
-release_clone="$clone_branch-release-tmp"
+release_clone="$orig_branch-release-tmp"
 
 echo "== Work tree will be in $release_clone"
 
 # Make a clone in a subdirectory and move there
 if ! [ -d "$release_clone" ]; then
     $VERBOSE "== Cloning to $release_clone"
-    git clone $git_quiet -b "$orig_branch" . "$release_clone"
+    git clone $git_quiet -b "$orig_branch" -o parent . "$release_clone"
```

Comment:
I'm not familiar with the details of the script, but I'm wondering whether it wouldn't be simpler to use worktrees instead of forking a separate repository. The (simplified) usage is similar to `git checkout`: you can specify an existing branch name as `<commit-ish>` or a new branch name using `-b <new-branch>`.


```
git worktree add [-b <new-branch>] <path> [<commit-ish>]
```

In your case, it would be

```
git worktree add $git_quiet -b "$orig_branch" "$release_clone"
```

Since the worktrees share all branches with the main repository, there is no need to push references from the clone to the parent.


**Side note**: I usually have several worktrees for the stable branches, wich I place next to the main repository:
```
git worktree add ../openssl-1.1.1 OpenSSL_1_1_1-stable
git worktree add ../openssl-1.0.2 OpenSSL_1_0_2-stable
```
But git has no problem if you place the worktree inside the repository.

Reply:
I would rather not for this.  The idea is that the script can do whatever it wants without affecting the parent repo, apart from the pushes.  In other words *if there's a mistake done somewhere*, it's more likely to be contained, and all you have to do is to rm the subdirectory and start over.

There are times when a little bit of paranoia is a good thing, you know? ;-)

Reply:
Oooooooo! I didn't know about worktrees! Nice!

Reply:
> There are times when a little bit of paranoia is a good thing, you know? ;-)

If you have your reasons for not using worktrees, that's ok for me. As I said, I'm not familiar with the details of the script and it was just a thought that came to my mind.

Reply:
@mattcaswell, `git worktree` came with git 2.5, in the summer of 2015.  News travel slow, I heard about it from @mspncp only a few months ago, and still haven't quite adopted it (I have a gazillion branches in my `master` checkout and am afraid of losing them).  I'm afraid that the shared nature of the worktrees will trip me some day...

Reply:
@levitte you'll love it once you get used to it. Because then you will have the same gazillion branches in all of your worktrees. ;-) And removing worktrees is absolutely safe if you stick to using `git worktree remove <worktree-path>`, because git won't let you remove the main working tree.

(Some people use `rm -rf <worktree-path>`, followed by `git worktree prune` in the main working tree, which is potentially dangerous.)

Reply:
Some more nice gimmicks you get with worktrees:  `git branch` will highlight the current branch of your worktree in green (prefixed by `*`), and the current branches of the other worktrees in cyan (prefixed by `+`). In very-verbose mode (`git branch -vv`) it will even show you the path of the worktrees:

```
msp@msppc:~/src/openssl$ git branch -vv  | grep '^[+*]'
+ OpenSSL-fips-2_0-stable                        d11e6a4410 (/home/msp/src/openssl-fips) [origin/OpenSSL-fips-2_0-stable] Add linux-mips32be target for new platform
+ ncp-1.1.1                                      4712166fc4 (/home/msp/src/openssl-1.1.1-ncp) x509: add missing X509 dup functions
+ ncp-1.0.2                                      60bf4243c5 (/home/msp/src/openssl-1.0.2-ncp) [ncp/ncp-1.0.2] EC_GROUP_new_from_ecparameters(): ignore unknown group errors in FIPS mode
+ romen-scmul_blinding_102                       0548c882ef (/home/msp/src/openssl-1.0.2) Implement blinding for EC scalar multiplication
+ pr-improve-reseed-on-fork-test-before-evp-rand 056a8fb1c6 (/home/msp/src/openssl-rof) [msp/pr-improve-reseed-on-fork-test-before-evp-rand] test/drbgtest: improve the reseed after fork test
+ OpenSSL_1_1_1-stable                           6328d3673f (/home/msp/src/openssl-1.1.1) [origin/OpenSSL_1_1_1-stable] Fix typos and repeated words
* pr-improve-reseed-on-fork-test                 b9774eaa62 [msp/pr-improve-reseed-on-fork-test] fixup! test/drbgtest: improve the reseed after fork test
```

The only 'limitation' is that you can never have the same branch checked out simultaneously in two different worktrees. But that limitation makes sense if you think about it, and it is not really a limitation, because you can always checkout a new topic branch at a given branch, if you really need to start from the same commit.

Reply:
> Because then you will have the same gazillion branches in all of your worktrees.

And that's probably the one thing that limits the usability for me...  I really don't want to see that gazillion *everywhere*.  But granted, that particular checkout could easily become a gazillion worktrees (one for each branch)...

Anyway, we digress.   For this script, I'm very convinced that letting it have a separate repo is the safer thing.